hazelcast/src/main/java/com/hazelcast/instance/DelegatingAddressPicker.java

```diff
@@ -0,0 +1,146 @@
+    private final BindAddressProvider bindAddressProvider;
+    private final NetworkConfig networkConfig;
+
+    private volatile InetSocketAddress bindAddress;
+    private volatile InetSocketAddress publicAddress;
+
+    private volatile ServerSocketChannel serverSocketChannel;
+
+    public DelegatingAddressPicker(BindAddressProvider bindAddressProvider, NetworkConfig networkConfig, ILogger logger) {
+        this.bindAddressProvider = bindAddressProvider;
+        this.networkConfig = networkConfig;
+        this.logger = logger;
+    }
+
+
+    @Override
+    public void pickAddress() throws Exception {
+        try {
+            bindAddress = bindAddressProvider.getBindAddress();
+            logger.info("Using bind address: " + publicAddress);
+
+            publicAddress = bindAddressProvider.getPublicAddress();
+            validatePublicAddress(publicAddress);
+            logger.info("Using public address: " + publicAddress);
+
+            serverSocketChannel = createServerSocketChannelInternal();
+
+            if (publicAddress.getPort() == 0) {
+                publicAddress = new InetSocketAddress(publicAddress.getAddress(), serverSocketChannel.socket().getLocalPort());
+            }
+        } catch (Exception e) {
+            logger.severe(e);
+            throw e;
+        }
+    }
+
+
+    private ServerSocketChannel createServerSocketChannelInternal() {
+        int portCount = networkConfig.getPortCount();
+        int port = bindAddress.getPort() == 0 ? networkConfig.getPort() : bindAddress.getPort();
+        boolean portAutoIncrement = networkConfig.isPortAutoIncrement();
+
+        int portTrialCount = port > 0 && portAutoIncrement ? portCount : 1;
+        if (port == 0) {
+            logger.info("No explicit port is given, system will pick up an ephemeral port.");
+        }
+
+        IOException error = null;
+        for (int i = 0; i < portTrialCount; i++) {
+            InetSocketAddress tmpBindAddress = new InetSocketAddress(bindAddress.getAddress(), port + i);
+            boolean reuseAddress = networkConfig.isReuseAddress();
+            logger.finest("inet reuseAddress:" + reuseAddress);
+
+            ServerSocket serverSocket = null;
+            ServerSocketChannel serverSocketChannel = null;
+            try {
+                serverSocketChannel = ServerSocketChannel.open();
+                serverSocket = serverSocketChannel.socket();
+                serverSocket.setReuseAddress(reuseAddress);
+                serverSocket.setSoTimeout(SOCKET_TIMEOUT_MILLIS);
+
+                logger.fine("Trying to bind inet socket address: " + tmpBindAddress);
+                serverSocket.bind(tmpBindAddress, SOCKET_BACKLOG_LENGTH);
+                logger.fine("Bind successful to inet socket address: " + serverSocket.getLocalSocketAddress());
+
+                //todo: ugly side-effect
+                bindAddress = tmpBindAddress;
+                return serverSocketChannel;
```

Comment:
`DefaultAddressPicker` configures socket channel as non-blocking. But this impl doesn't as far as I see.
```
serverSocketChannel.configureBlocking(false);
```

Reply:
The `DefaultAddressPicker` also respects the `GroupProperty#SOCKET_SERVER_BIND_ANY` property. Should this be added as well?